api/wasm/cpp/proxy_wasm_externs.h

```diff
@@ -11,11 +11,14 @@
 /*
    API Calls into the VM.
 
-   extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onConfigure(char* configuration, int size);
-   extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onStart();
-   extern "C" EMSCRIPTEN_KEEPALIVE int main();  // only called if proxy_onStart() is not available.
-   extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onTick();
-   extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onCreate(uint32_t context_id);
+
+   // Non-stream calls.
+   extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onStart(uint32_t root_context_id, uint32_t root_id_ptr, uint32_t root_id_size);
```

Comment:
Why do we need both: `root_context_id` (number) and `root_id` (string)?

Reply:
Yes, because the root_context_id is arbitrary and opaque and is selected by the underlying system whereas the root_id is selected by the by wasm programmer in the configuration.  In particular, the root_context_id depends on how many root_ids there are and in what order they are initialized.  The root_id is converted into a root_context_id which makes the dispatch code faster since it doesn't need to hash the string and it makes the ABI more consistent since the context is always uniquely defined by the context_id independent of the type of handler.

Reply:
If it's not assigned by the WASM programer but the host, shouldn't `root_context_id` be a return value, then?

Reply:
This is a call into the VM, so the host assigned it outside this call and then passed it in.